### ISSUE TITLE
Typing ':' (colon) after open-paren should not commit completion

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#13527]
If you type "i" in `($$)`, `int` is only offered as soft-completion. But it should offer `int` as a hard-selection instead, but with ':' (colon) not being a commit character.

```VB

        <WpfFact, Trait(Traits.Feature, Traits.Features.Completion)>
        Public Async Function TestDeconstructionDeclaration4() As Task
            Using state = TestState.CreateCSharpTestState(
                  <Document><![CDATA[
class C
{
    public void Foo()
    {
        ($$)
    }
}]]></Document>)

                state.SendTypeChars("i")
                Await state.WaitForAsynchronousOperationsAsync()
                Await state.AssertSelectedCompletionItem(displayText:="int", isSoftSelected:=True)
            End Using
        End Function
```

FYI for @CyrusNajmabadi

**Comment by CyrusNajmabadi** [comment]
Thanks.

**Comment by jcouv** [comment]
@CyrusNajmabadi I'm wondering how aggressive the change should be.
For `KeywordCompletionProvider` and `SymbolCompletionProvider` in C#, I'm tempted to remove ':' as a commit character in all cases. What would be the downside of that?

Also, I looked around `CSharpCompletionCommandHandlerTests` for verification of commit characters, but couldn't find anything handy. Should I add some validation helpers there?
Alternatively, I found that `OverrideCompletionProviderTests` verify commit characters (using  `VerifyCustomCommitProviderAsync`), but that seems to focus on a single completion provider. I could do that too.
Which approach do you recommend?

**Comment by CyrusNajmabadi** [comment]
> For KeywordCompletionProvider and SymbolCompletionProvider in C#, I'm tempted to remove ':' as a commit character in all cases. What would be the downside of that?

That would not be desirable.  For example, if someone wrote "case Color.Bl<colon>" then it would not complete out to "Black".

> Also, I looked around CSharpCompletionCommandHandlerTests for verification of commit characters, but couldn't find anything handy. Should I add some validation helpers there?

No.  I would just write two tests one that verifies that colon does not cause commit in a parenthesized expression case.  And one that verifies that it does with something like "case X.Y".

> Alternatively, I found that OverrideCompletionProviderTests verify commit characters (using VerifyCustomCommitProviderAsync), but that seems to focus on a single completion provider. I could do that too.

Do not do this.  These tests are completely bogus and cannot be trusted.  They are attempting to validate things by testing against hte API.  But they're not actually verifying the experience.  Anything related to how typing affects completion should be done though the XXXCommandHandlerTests.

**Comment by jcouv** [comment]
@CyrusNajmabadi 
Thanks for the clarifications.

I think I have it working for `SymbolCompletionProvider` (see PR https://github.com/dotnet/roslyn/pull/13638). 

For `KeywordCompletionProvider` I think I'll have to build a `SyntaxContext`, so that the completion provider can tell which rule it should produce depending on context. Does that sound alright?

**Comment by CyrusNajmabadi** [comment]
doesn't KeywordCompletionProvider already use a SyntaxContext?

**Comment by jcouv** [comment]
Fixed by #13638.
